File: backend/src/services/auth_service.py
from sqlalchemy.orm import Session
from datetime import datetime, timedelta
from typing import Optional
from ..models.user import User, PasswordResetToken
from ..schemas.auth import LoginRequest, RegisterRequest
from ..utils.security import (
    hash_password,
    verify_password,
    create_access_token,
    create_refresh_token,
    generate_password_reset_token,
)
from ..utils.email import send_password_reset_email
from fastapi import HTTPException, status


class AuthService:
    """Service for authentication operations."""
    
    @staticmethod
    def authenticate_user(db: Session, username: str, password: str) -> Optional[User]:
        """
        Authenticate a user with username and password.
        
        Args:
            db: Database session
            username: Username
            password: Plain text password
            
        Returns:
            User object if authentication successful, None otherwise
        """
        user = db.query(User).filter(
            User.username == username,
            User.is_active == True
        ).first()
        
        if not user:
            return None
        
        if not verify_password(password, user.hashed_password):
            return None
        
        # Update last login
        user.last_login = datetime.utcnow()
        db.commit()
        
        return user
    
    @staticmethod
    def login(db: Session, login_data: LoginRequest) -> tuple[User, str, str]:
        """
        Login a user and generate tokens.
        
        Args:
            db: Database session
            login_data: Login credentials
            
        Returns:
            Tuple of (user, access_token, refresh_token)
            
        Raises:
            HTTPException: If authentication fails
        """
        user = AuthService.authenticate_user(
            db, login_data.username, login_data.password
        )
        
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect username or password",
            )
        
        # Generate tokens
        token_data = {
            "sub": str(user.id),
            "username": user.username,
            "role": user.role.value,
        }
        access_token = create_access_token(token_data)
        refresh_token = create_refresh_token(token_data)
        
        return user, access_token, refresh_token
    
    @staticmethod
    def register(db: Session, register_data: RegisterRequest) -> User:
        """
        Register a new user.
        
        Args:
            db: Database session
            register_data: Registration data
            
        Returns:
            Created user object
            
        Raises:
            HTTPException: If username or email already exists
        """
        # Check if username exists
        existing_user = db.query(User).filter(
            User.username == register_data.username
        ).first()
        if existing_user:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username already registered",
            )
        
        # Check if email exists
        existing_email = db.query(User).filter(
            User.email == register_data.email
        ).first()
        if existing_email:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already registered",
            )
        
        # Create new user
        hashed_pw = hash_password(register_data.password)
        new_user = User(
            username=register_data.username,
            email=register_data.email,
            hashed_password=hashed_pw,
            full_name=register_data.full_name,
        )
        
        db.add(new_user)
        db.commit()
        db.refresh(new_user)
        
        return new_user
    
    @staticmethod
    def forgot_password(db: Session, email: str) -> bool:
        """
        Initiate password reset process.
        
        Args:
            db: Database session
            email: User email
            
        Returns:
            True if email sent successfully
        """
        user = db.query(User).filter(
            User.email == email,
            User.is_active == True
        ).first()
        
        if not user:
            # Don't reveal if email exists
            return True
        
        # Generate reset token
        token = generate_password_reset_token()
        expires_at = datetime.utcnow() + timedelta(hours=24)
        
        # Save token to database
        reset_token = PasswordResetToken(
            user_id=user.id,
            token=token,
            expires_at=expires_at,
        )
        db.add(reset_token)
        db.commit()
        
        # Send email
        send_password_reset_email(user.email, token)
        
        return True
    
    @staticmethod
    def reset_password(db: Session, token: str, new_password: str) -> bool:
        """
        Reset password using token.
        
        Args:
            db: Database session
            token: Password reset token
            new_password: New password
            
        Returns:
            True if password reset successful
            
        Raises:
            HTTPException: If token is invalid or expired
        """
        reset_token = db.query(PasswordResetToken).filter(
            PasswordResetToken.token == token
        ).first()
        
        if not reset_token:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid reset token",
            )
        
        if not reset_token.is_valid():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Reset token has expired or been used",
            )
        
        # Get user
        user = db.query(User).filter(User.id == reset_token.user_id).first()
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        
        # Update password
        user.hashed_password = hash_password(new_password)
        reset_token.is_used = True
        
        db.commit()
        
        return True
    
    @staticmethod
    def change_password(
        db: Session, user_id: int, current_password: str, new_password: str
    ) -> bool:
        """
        Change user password.
        
        Args:
            db: Database session
            user_id: User ID
            current_password: Current password
            new_password: New password
            
        Returns:
            True if password changed successfully
            
        Raises:
            HTTPException: If current password is incorrect
        """
        user = db.query(User).filter(User.id == user_id).first()
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )
        
        if not verify_password(current_password, user.hashed_password):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Current password is incorrect",
            )
        
        user.hashed_password = hash_password(new_password)
        db.commit()
        
        return True

File: go-backend/internal/services/auth_service.go
package services

import (
	"errors"
	"fmt"
	"time"

	"github.com/estate-docs/go-backend/internal/config"
	"github.com/estate-docs/go-backend/internal/models"
	"github.com/estate-docs/go-backend/internal/utils"
	"gorm.io/gorm"
)

type AuthService struct {
	DB  *gorm.DB
	Cfg *config.Config
}

func NewAuthService(db *gorm.DB, cfg *config.Config) *AuthService {
	return &AuthService{DB: db, Cfg: cfg}
}

type RegisterInput struct {
	Username string `json:"username"`
	Email    string `json:"email"`
	Password string `json:"password"`
	FullName string `json:"full_name"`
}

type LoginInput struct {
	Username string `json:"username"`
	Password string `json:"password"`
}

func (s *AuthService) Register(input RegisterInput) (*models.User, error) {
	// Check existing username
	var count int64
	s.DB.Model(&models.User{}).Where("username = ?", input.Username).Count(&count)
	if count > 0 {
		return nil, errors.New("username already exists")
	}

	// Check existing email
	s.DB.Model(&models.User{}).Where("email = ?", input.Email).Count(&count)
	if count > 0 {
		return nil, errors.New("email already exists")
	}

	hashed, err := utils.HashPassword(input.Password)
	if err != nil {
		return nil, fmt.Errorf("failed to hash password: %w", err)
	}

	user := models.User{
		Username:       input.Username,
		Email:          input.Email,
		HashedPassword: hashed,
		Role:           models.RoleUser,
		IsActive:       true,
		CreatedAt:      time.Now().UTC(),
		UpdatedAt:      time.Now().UTC(),
	}
	if input.FullName != "" {
		user.FullName = &input.FullName
	}

	if err := s.DB.Create(&user).Error; err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}
	return &user, nil
}

// Login returns (user, accessToken, refreshToken, error)
func (s *AuthService) Login(input LoginInput) (*models.User, string, string, error) {
	var user models.User
	if err := s.DB.Where("username = ? AND is_active = true", input.Username).First(&user).Error; err != nil {
		return nil, "", "", errors.New("invalid username or password")
	}

	if !utils.VerifyPassword(input.Password, user.HashedPassword) {
		return nil, "", "", errors.New("invalid username or password")
	}

	// Update last login
	now := time.Now().UTC()
	user.LastLogin = &now
	s.DB.Save(&user)

	tokenData := map[string]interface{}{
		"sub":      fmt.Sprintf("%d", user.ID),
		"username": user.Username,
		"role":     string(user.Role),
	}

	accessToken, err := utils.CreateAccessToken(s.Cfg, tokenData)
	if err != nil {
		return nil, "", "", fmt.Errorf("failed to create access token: %w", err)
	}

	refreshJWT, jti, expiresAt, err := utils.CreateRefreshToken(s.Cfg, tokenData)
	if err != nil {
		return nil, "", "", fmt.Errorf("failed to create refresh token: %w", err)
	}

	// Store refresh token server-side
	rt := models.RefreshToken{
		UserID:    user.ID,
		TokenJTI:  jti,
		ExpiresAt: expiresAt,
		CreatedAt: time.Now().UTC(),
		UpdatedAt: time.Now().UTC(),
	}
	s.DB.Create(&rt)

	return &user, accessToken, refreshJWT, nil
}

func (s *AuthService) ForgotPassword(email string) error {
	var user models.User
	if err := s.DB.Where("email = ?", email).First(&user).Error; err != nil {
		// Don't reveal whether email exists
		return nil
	}

	token := utils.GeneratePasswordResetToken()
	prt := models.PasswordResetToken{
		UserID:    user.ID,
		Token:     token,
		ExpiresAt: time.Now().UTC().Add(1 * time.Hour),
		CreatedAt: time.Now().UTC(),
		UpdatedAt: time.Now().UTC(),
	}
	s.DB.Create(&prt)

	// TODO: Send email with reset token
	return nil
}

func (s *AuthService) ResetPassword(token, newPassword string) error {
	var prt models.PasswordResetToken
	if err := s.DB.Where("token = ?", token).First(&prt).Error; err != nil {
		return errors.New("invalid or expired reset token")
	}

	if !prt.IsValid() {
		return errors.New("invalid or expired reset token")
	}

	hashed, err := utils.HashPassword(newPassword)
	if err != nil {
		return fmt.Errorf("failed to hash password: %w", err)
	}

	// Update password
	s.DB.Model(&models.User{}).Where("id = ?", prt.UserID).Update("hashed_password", hashed)

	// Mark token as used
	prt.IsUsed = true
	s.DB.Save(&prt)

	return nil
}

func (s *AuthService) ChangePassword(userID int, currentPassword, newPassword string) error {
	var user models.User
	if err := s.DB.First(&user, userID).Error; err != nil {
		return errors.New("user not found")
	}

	if !utils.VerifyPassword(currentPassword, user.HashedPassword) {
		return errors.New("current password is incorrect")
	}

	hashed, err := utils.HashPassword(newPassword)
	if err != nil {
		return fmt.Errorf("failed to hash password: %w", err)
	}

	user.HashedPassword = hashed
	user.UpdatedAt = time.Now().UTC()
	return s.DB.Save(&user).Error
}
